src/process.c

```diff
@@ -111,6 +113,153 @@ char * rcutils_get_executable_name(rcutils_allocator_t allocator)
+  if (NULL == process) {
+    return NULL;
+  }
+  process->allocator = *allocator;
+
+#if defined _WIN32 || defined __CYGWIN__
+  LPSTR cmd = rcutils_join(args, " ", *allocator);
+  if (NULL == cmd) {
+    rcutils_process_close(process);
+    return NULL;
+  }
+
+  STARTUPINFO start_info = {sizeof(start_info)};
+  PROCESS_INFORMATION process_info = {0};
+  if (!CreateProcess(
+          NULL, cmd, NULL, NULL, TRUE, 0,
+          NULL, NULL, &start_info, &process_info))
+  {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to create process: %lu", GetLastError());
+    allocator->deallocate(cmd, &allocator->state);
+    rcutils_process_close(process);
+    return NULL;
+  }
+
+  allocator->deallocate(cmd, &allocator->state);
+
+  CloseHandle(process_info.hThread);
+
+  process->handle = process_info.hProcess;
+  process->pid = process_info.dwProcessId;
+
+  return process;
+#else
+  rcutils_string_array_t argv = rcutils_get_zero_initialized_string_array();
+  if (RCUTILS_RET_OK != rcutils_string_array_init(&argv, args->size + 1, allocator)) {
+    return NULL;
+  }
+
+  const char * cmd = args->data[0];
+  for(size_t i = 0; i < args->size; i++) {
+    argv.data[i] = args->data[i];
+  }
+
+  process->pid = fork();
+  if (-1 == process->pid) {
+    int error = errno;
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to fork process: %d", error);
+    rcutils_process_close(process);
+    return NULL;
+  } else if (0 != process->pid) {
+    return process;
+  }
+
+  int exec_ret = execvp(cmd, argv.data);
+
+  RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING("Failed to execute process: %d", exec_ret);
+
+  for(size_t i = 0; i < argv.size; i++) {
+    argv.data[i] = NULL;
+  }
+
+  rcutils_ret_t rcutils_ret = rcutils_string_array_fini(&argv);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING(
+      "Failed to deinitialize argv: %d", rcutils_ret);
+  }
+
+  exit(127);
+#endif
+}
+
+void
+rcutils_process_close(rcutils_process_t * process)
+{
+  if (NULL == process) {
+    return;
+  }
+
+  rcutils_allocator_t allocator = process->allocator;
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
+    &allocator, "allocator is invalid", return );
+
+#if defined _WIN32 || defined __CYGWIN__
+  CloseHandle(process->handle);
+#endif
+
+  allocator.deallocate(process, allocator.state);
+}
+
+rcutils_ret_t
+rcutils_process_wait(const rcutils_process_t * process, int *exit_code)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(process, RCUTILS_RET_INVALID_ARGUMENT);
+
+#if defined _WIN32 || defined __CYGWIN__
+  DWORD status;
+
+  if (WAIT_FAILED == WaitForSingleObject(process->handle, INFINITE)) {
+    return RCUTILS_RET_ERROR;
+  }
+
+  if (NULL != exit_code) {
+    if (!GetExitCodeProcess(process->handle, &status)) {
+      return RCUTILS_RET_ERROR;
+    }
+    *exit_code = status;
+  }
+#else
+  int status;
+
+  int ret = waitpid(process->pid, &status, 0);
+  if (-1 == ret) {
+    return RCUTILS_RET_ERROR;
```

Comment:
```suggestion
    int error = errno;
    RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING("Failed to wait for process %d: %d (%s)", process->pid, error, strerror(error));
    return RCUTILS_RET_ERROR;
```

Reply:
Missed this one, but I just added it in 13872ec126b6.